### PR TITLE
Add Notification service to send messages to the queue for the rejected envelopes

### PR DESCRIPTION
**Pull request by aliveni**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/BPS-1096

### Change description ###
Add Notification service
- To send messages to notifications queue for the rejected envelopes
-  Add an event when notification message is sent

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/blobrouter/services/NotificationService.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+    public void sendNotifications() {
+        List<RejectedEnvelope> envelopes = rejectedEnvelopeRepository.getRejectedEnvelopes();
+        envelopes.forEach(
+            env -> {
+                log.info(
+                    "Send message to notifications queue. File name: {} Container: {}", env.fileName, env.container
+                );
+                notificationsPublisher.publish(mapToNotificationMessage(env));
+                envelopeService.saveEvent(env.envelopeId, NOTIFICATION_SENT);
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/services/NotificationService.java]: is that enought? will this prevent sending a notification for the same envelope again?

File: src/test/java/uk/gov/hmcts/reform/blobrouter/services/NotificationServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+        // when
+        service.sendNotifications();
+
+        // then
+        verify(rejectedEnvelopeRepository).getRejectedEnvelopes();
+        verify(notificationsPublisher, times(2)).publish(any(NotificationMsg.class));
+        verify(envelopeService).saveEvent(envelopeId1, EventType.NOTIFICATION_SENT);
+        verify(envelopeService).saveEvent(envelopeId1, EventType.NOTIFICATION_SENT);
+    }
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/blobrouter/services/NotificationServiceTest.java]: an integration test would be nice, to check that notification gets sent only once (ie, second call of the service will ignore envelopes for which notification has already been sent)

File: src/main/java/uk/gov/hmcts/reform/blobrouter/services/NotificationService.java (hunk 1 above)
Comment (aliveni) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/services/NotificationService.java]: No, I'll have to change the query to the envelopes with event `NOTIFICATION_SENT`

File: src/main/java/uk/gov/hmcts/reform/blobrouter/services/NotificationService.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+import java.util.List;
+
+import static uk.gov.hmcts.reform.blobrouter.data.events.EventType.NOTIFICATION_SENT;
+
+public class NotificationService {
+
+    private static final Logger log = LoggerFactory.getLogger(NotificationService.class);
+
+    private static final String SERVICE_NAME = "Blob_Router";
```
[the remark this answers is not in the document]
Comment (aliveni) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/services/NotificationService.java]: updated

**Comment by aliveni** [comment]
> missing integration test

Added the integration tests

File: src/main/java/uk/gov/hmcts/reform/blobrouter/services/NotificationService.java (hunk 1 above)
Comment (aliveni) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/services/NotificationService.java]: Updated the query and added additional tests to `RejectedEnvelopeRepositoryTest`

File: src/main/java/uk/gov/hmcts/reform/blobrouter/data/rejectedenvelope/RejectedEnvelopeRepository.java [hunk 4]
```diff
@@ -21,10 +21,24 @@ public RejectedEnvelopeRepository(
...
+                + "      (CASE WHEN type = 'NOTIFICATION_SENT' THEN 1 else 0 END) AS notification_event "
+                + "      FROM envelope_events "
+                + "      WHERE type IN ('REJECTED', 'NOTIFICATION_SENT') "
+                + "      GROUP BY envelope_id, type "
+                + "     ) events " /* envelopes having Rejected and Notification_sent events */
+                + "  GROUP BY envelope_id "
+                + " ) rejected_envelopes " /* rejected envelope_ids from events */
+                + " ON e.envelope_id = rejected_envelopes.envelope_id " /* join on events and rejected envelopes */
+                + " WHERE e.type = 'REJECTED' "
+                + "     AND rejected_envelopes.notification_sent_count = 0", /* filter notification_sent */
```
[the remark this answers is not in the document]
Comment (aliveni) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/data/rejectedenvelope/RejectedEnvelopeRepository.java]: Updated the format. Also, simplified the query

File: src/main/java/uk/gov/hmcts/reform/blobrouter/data/rejectedenvelope/RejectedEnvelopeRepository.java [hunk 5]
```diff
@@ -21,10 +21,22 @@ public RejectedEnvelopeRepository(
...
+                + "     FROM "
+                + "         (SELECT envelope_id, type "
+                + "         FROM envelope_events "
+                + "         WHERE type IN ('REJECTED', 'NOTIFICATION_SENT') "
+                + "     ) events " /* Rejected and Notification_sent events */
+                + "     GROUP BY envelope_id "
+                + " ) rejected_envelopes " /* envelope_ids and events_count which would be 2 if notification_sent */
+                + "     ON e.envelope_id = rejected_envelopes.envelope_id "
+                + " WHERE e.type = 'REJECTED' "
+                + "     AND rejected_envelopes.events_count = 1", /* filter notification_sent events */
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/data/rejectedenvelope/RejectedEnvelopeRepository.java:39]: I thought we are going to add a new column to envelope to see if notification was sent or not ? Query looks complex to me

Comment (aliveni) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/data/rejectedenvelope/RejectedEnvelopeRepository.java]: updated the query to use the `pending_notification` column in envelope

File: src/main/java/uk/gov/hmcts/reform/blobrouter/services/NotificationService.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        EnvelopeService envelopeService
+    ) {
+        this.notificationsPublisher = notificationsPublisher;
+        this.rejectedEnvelopeRepository = rejectedEnvelopeRepository;
+        this.envelopeService = envelopeService;
+    }
+
+    public void sendNotifications() {
+        List<RejectedEnvelope> envelopes = rejectedEnvelopeRepository.getRejectedEnvelopes();
+        envelopes.forEach(
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/services/NotificationService.java:39]: we can probably get rid of local var here just chain the methods

File: src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/services/NotificationServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+        envelopeService.saveEvent(envelopeId3, DELETED);
+
+        // when
+        notificationService.sendNotifications();
+
+        // then
+        verify(notificationsPublisher, times(2)).publish(any());
+
+        Optional<Envelope> envelope1 = envelopeService.findEnvelope(envelopeId1);
+        assertThat(envelope1).hasValueSatisfying(env -> assertThat(env.pendingNotification).isEqualTo(false));
```
Comment (doncem) [inline src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/services/NotificationServiceTest.java:92]: usually it's `assertThat(someBoolean).isFalse()`. there's another one on line 97

File: src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/services/NotificationServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+        envelopeService.markAsDispatched(envelopeId2);
+        envelopeService.saveEvent(envelopeId1, DELETED);
+
+        // rejected but notification NOT sent
+        var envelopeId3 = envelopeService.createNewEnvelope("bulkscan", "blob3.zip", now());
+        envelopeService.markAsRejected(envelopeId3, "invalid signature");
+        envelopeService.saveEvent(envelopeId3, DELETED);
+
+        // when
+        notificationService.sendNotifications();
```
Comment (rkondratowicz) [inline src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/services/NotificationServiceTest.java:122]: how about we call the service twice?
first time it should send notifications, the second time it shouldn't

**Pull request merged**